### PR TITLE
Refactoring text segment and word completion logic.

### PR DESCRIPTION
**Pull request by filipi87**
## Summary

Fixes a text truncation bug in TTS word completion tracking, and rewrites how raw TTS word-timestamp tokens are matched against the TTS text, which is what made the fix tractable to implement cleanly.

## Fix

- `strip_markup()`'s behavior — "an unclosed `<` swallows the rest of the string" — is correct when parsing a *streamed* word-timestamp fragment that may genuinely be mid-tag (e.g. a multi-attribute SSML tag some TTS providers split across several tokens). But it was also being used to build the default `user_facing_text` from a *complete* `tts_text`, and to compute `TextSegment.is_transformed`. In a complete string, a lone `<` is just
  content, not a truncated tag — so TTS output like `"5 < 10"` or an emoticon like `"<3"` had everything after the `<` silently dropped from the user-facing text and LLM context.

- Added `strip_complete_markup()` — a paired-tag-only stripper for complete/static text — and switched those two call sites to it. `strip_markup()` keeps its truncating behavior for the one place it's actually correct: matching a raw,   possibly-mid-tag word-timestamp token against the remaining TTS text.

## Internal refactor

- `TextSegmentMap` maps a single raw TTS cursor across three parallel texts — `tts_text`, `original_text` (user-facing), and `llm_text` — and this part is unchanged: converting a consumed TTS span into positions in `original_text` and `llm_text` is still done by counting normalized alphanumeric characters and walking that many alnum chars forward via `advance_by_alnums`.

- What changed is the step *before* that: how an incoming raw word-timestamp token is matched against the TTS text to determine how far the raw cursor moves within the current segment. Previously, the caller's word was normalized and consumed against the segment's alnum *budget* (`tts_alnum_count`), without the matching logic ever looking at the segment's actual raw characters.

- Now, the matching engine was rewritten to work directly on raw text instead of char counts:
  - `_classify_hop()` matches an incoming word against a segment's *remaining raw text* using three strategies in order — literal match, whitespace-stripped match, markup-stripped match — and classifies the result as:
    - PLACED # word fits within this segment; stop here
    - CROSSES # word runs past this segment; drain it and carry the remainder
    - EXHAUSTED # no spoken content left here; drain it, keep the whole word
    - NO_MATCH # word doesn't belong here; nudge past leading punctuation, stop 
  - `_advance_raw()` / `_word_matches_remaining()` walk segments using that classification, replacing the old alnum-count state (`_seg_consumed`, `_received`, `_tts_normalized`) with a single raw cursor (`_seg_raw_pos`,  exposed as `raw_pos`).
  - `WordCompletionTracker` now delegates completion/cursor tracking almost entirely to `TextSegmentMap` (`is_complete`, `raw_pos`, `last_overflow`) instead of maintaining a parallel char-count state of its own.

>  Matching against actual raw text (rather than an alnum budget) is what makes the fix straightforward: it can tell a mid-tag fragment from a complete text's stray `<` directly, instead of requiring more char-count special-casing. 

> This is purely an internal implementation change — no public API or behavior change for existing callers.

## Test plan
  - [x] `uv run pytest tests/test_text_segment_map.py tests/test_word_completion_tracker.py` — 212 passed
  - [x] Added regression tests for the stray-`<` truncation fix (`strip_complete_markup`, `TextSegmentMap` segment
  classification, `WordCompletionTracker` default `user_facing_text`)
  - [x] Existing markup/SSML-fragment, symbol-substitution, and CJK tests still pass unchanged, confirming no behavior drift from the refactor

### CODE REVIEW

File: tests/test_text_segment_map.py [hunk 1]
```diff
@@ -6,7 +6,66 @@
...
+    def test_raw_len_beyond_available_returns_full_length(self):
+        self.assertEqual(_raw_len_for_clean_chars("<b>x</b>", 99), len("<b>x</b>"))
+
+    def test_raw_len_agrees_with_strip_markup(self):
+        # Consuming len(strip_markup(t)) clean chars must land at/after the last
+        # clean char for any input.
+        for t in ["<speak>hello</speak>", "1234-5678", "<a>x</a><b>y</b>", "plain"]:
+            clean = strip_markup(t)
+            pos = _raw_len_for_clean_chars(t, len(clean))
+            self.assertEqual(strip_markup(t[:pos]), clean)
```
Comment (mattieruth) [inline tests/test_text_segment_map.py:60]: i'm a little confused how this assert is doing the check for "landing at/after the last clean char".  i guess it checks that there's nothing extra in the string up to the end of the clean characters, but maybe we should check, too that the next character either doesn't exist or is "<"?

File: tests/test_word_completion_tracker.py [hunk 2]
```diff
@@ -1550,11 +1551,25 @@ def test_reset_clears_user_facing_pos(self):
         self.assertEqual(tracker.get_remaining_user_facing_text(), self.USER_FACING_TEXT)
 
 
+class TestUserFacingTextStrayAngleBracket(unittest.TestCase):
+    """A literal '<' with no matching '>' in tts_text (e.g. an emoticon like "<3")
+    is real content, not a truncated SSML tag. When user_facing_text is omitted,
+    the default derived from tts_text must not silently drop everything after it."""
+
+    def test_default_user_facing_text_keeps_text_after_stray_bracket(self):
```
Comment (mattieruth) [inline tests/test_word_completion_tracker.py:1576]: Looks like the markdown version of this is part of the TestWordCompletionTrackerLLMText? but those don't always test the tracker.iscomplete ... maybe just double check that we are well covered with those tests.

**Review by mattieruth — APPROVED** [review]
I'm understanding this at the high level and at the very low-level but struggling to kind of understand it all together. I think I have to rely on the unit tests to catch things so I ended up focusing mostly there.  I'm happy to approve but maybe another look from paul would be good before merging.

**Comment by kompfner** [comment]
> is correct when parsing a streamed word-timestamp fragment that may genuinely be mid-tag

But couldn't the streamed text just as easily also be in the middle of outputting something containing "<3"? 🤔 (So, wouldn't `strip_markup()` make the same mistake then?)

I guess with this fix we're _at least_ fixing it for the case where the whole response is something like "I <3 ice cream", so strictly better than before.

**Comment by kompfner** [comment]
> maybe another look from paul

😬

> I have to rely on the unit tests to catch things so I ended up focusing mostly there

Honestly, this stuff is so very complex and fiddly that that approach sounds like the only way in without spending an inordinate amount of time. I don't have the full picture either.

**Comment by filipi87** [comment]
>> is correct when parsing a streamed word-timestamp fragment that may genuinely be mid-tag

> But couldn't the streamed text just as easily also be in the middle of outputting something containing "<3"? 🤔 (So, wouldn't strip_markup() make the same mistake then?)

Hi @kompfner, `_classify_hop` tries literal matching first, and `"<3"` matches the raw segment text character for character without ever touching `strip_markup`, which is why it usually works in this case.

But you're right, there could still be a rare edge case where this happens. I'm not sure. I will try add some tests for this.

File: tests/test_word_completion_tracker.py [hunk 3]
```diff
@@ -1550,11 +1551,49 @@ def test_reset_clears_user_facing_pos(self):
...
+            tracker.add_word_and_check_complete(word)
+
+        self.assertTrue(tracker.word_belongs_here("<3"))
+        complete = tracker.add_word_and_check_complete("<3")
+        self.assertFalse(complete)
+        self.assertIsNone(tracker.get_overflow_word())
+
+        complete = tracker.add_word_and_check_complete("always")
+        self.assertTrue(complete)
+        self.assertEqual(tracker.get_accumulated_user_facing_text(), text)
```
Comment (filipi87) [inline tests/test_word_completion_tracker.py:1606]: I have added this test, to show it is working as expected in this case.

File: tests/test_text_segment_map.py (hunk 1 above)
Comment (filipi87) [reply tests/test_text_segment_map.py:60]: Yep, I think you are right. I have pushed a fix improving the test.

File: tests/test_word_completion_tracker.py (hunk 2 above)
Comment (filipi87) [reply tests/test_word_completion_tracker.py:1576]: I have improved the tests, to where it makes sense, test if it is complete or not. 🙂
